### PR TITLE
Support filtering MoneyAddresses

### PR DESCRIPTION
**Pull request by tomdaffurn**
Add:
- `MoneyAddressFilter`: a function that tests a `MoneyAddress` against arbitrary conditions, and returns an arbitrary type
- `Filter`: a function that accepts an slice of `MoneyAddress`es and a `MoneyAddressFilter`, and returns only the maddrs that match the filter

### CODE REVIEW

File: maddr/mobilemoney_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package maddr
```
Comment (mistermoe) [inline maddr/mobilemoney_test.go]: let's use blackbox testing so we can see what the api surface looks like in the tests. `package maddr_test` oughta do it

File: maddr/maddr.go [hunk 2]
```diff
@@ -26,28 +26,32 @@ func FromDIDService(svc didcore.Service) ([]MoneyAddress, error) {
...
-		maddrs[i] = maddr
+	delimIDX := strings.IndexRune(urn.NSS, ':')
+	if delimIDX == -1 {
+		return nil, fmt.Errorf("invalid money address. expected urn:[currency]:[protocol]:[pss]. got %s", maddr)
 	}
 
-	return maddrs, nil
+	return &MoneyAddress{
+		URN:      urn,
+		ID:       serviceID,
```
Comment (mistermoe) [inline maddr/maddr.go:52]: we should create an issue to remove `ID`. feels odd to have it as part of the `MoneyAddress`. it's a did document service concept

File: maddr/mobilemoney_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+			}
+		})
+	}
+}
+
+func newMoneyAddress(t *testing.T, maddr string) *MoneyAddress {
+	m, err := FromURN("didpay", maddr)
+	assert.NoError(t, err)
+	return m
+}
```
Comment (mistermoe) [inline maddr/mobilemoney_test.go]: thoughts on including a `MustParse` as part of the lib itself. We see it i a lot of the other go libs we use e.g. `decimal`, `uuid`, `typeid`. Prevents us from having to write functions like this.

File: dap/filter_test.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+	out, err := Filter(maddrs, maddr.KESMoneyAddressFilter)
+
+	assert.NoError(t, err)
+	assert.NotNil(t, out)
+	assert.Len(t, out, 1)
+	actualMaddr := out[0]
+	if !reflect.DeepEqual(actualMaddr, expectedOut) {
+		t.Errorf("Filter() got = %v, want %v", expectedOut, actualMaddr)
+	}
```
Comment (mistermoe) [inline dap/filter_test.go]: interesting. does `assert` not have a way to do this? i think `Equal` and/or `Compare` might provide deep equality checks. @alecthomas keep me honest here

File: dap/filter.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import "github.com/TBD54566975/dap-go/maddr"
+
+// MoneyAddressFilter accepts a maddr.MoneyAddress, tests if it matches the filter condition, and returns either a more
+// specific money address type, or nil
+type MoneyAddressFilter[T any] func(maddr maddr.MoneyAddress) (*T, error)
+
+// Filter applies the MoneyAddressFilter to each element in maddrs, and returns a slice of the MoneyAddresses that
+// matched
+func Filter[T any](maddrs []maddr.MoneyAddress, filter MoneyAddressFilter[T]) ([]T, error) {
```
Comment (mistermoe) [inline dap/filter.go]: putting `Filter` in the `maddr` package would likely provide more context as to what this is a filter for without having to look at the argument types. `maddr.Filter` vs. `dap.Filter`. the latter makes me think i'm filtering daps

File: dap/filter_test.go (hunk 4 above)
Comment (alecthomas) [reply dap/filter_test.go]: Yep it does do a deep comparison, and it will output a diff.

Comment (tomdaffurn) [reply dap/filter_test.go]: Thanks! This was the codegened test format that IntelliJ makes

File: maddr/maddr.go (hunk 2 above)
Comment (tomdaffurn) [reply maddr/maddr.go:52]: Just drop it entirely?

File: maddr/maddr.go [hunk 6]
```diff
@@ -26,28 +26,40 @@ func FromDIDService(svc didcore.Service) ([]MoneyAddress, error) {
...
+		Currency: urn.NID,
+		Protocol: urn.NSS[:delimIDX],
+		PSS:      urn.NSS[delimIDX+1:],
+	}, nil
+}
+
+func MustParse(serviceID string, maddrURN string) MoneyAddress {
+	maddr, err := Parse(serviceID, maddrURN)
+	if err != nil {
+		panic(`maddr: Parse(` + maddrURN + `): ` + err.Error())
```
Comment (mistermoe) [inline maddr/maddr.go:62]: nice panic message!

**Review by mistermoe — APPROVED** [review]

**Pull request merged**